src/checkqueue.h

```diff
@@ -167,15 +167,20 @@ class CCheckQueue
     //! Add a batch of checks to the queue
     void Add(std::vector<T>& vChecks)
     {
-        LOCK(m_mutex);
-        for (T& check : vChecks) {
-            queue.push_back(T());
-            check.swap(queue.back());
+        unsigned int num_of_new_checks;
+        {
+            LOCK(m_mutex);
+            for (T& check : vChecks) {
+                queue.push_back(T());
+                check.swap(queue.back());
+            }
+            num_of_new_checks = vChecks.size();
```

Comment:
nit: I think you could also write `const auto num_of_new_checks = vChecks.size();` above, saves one line of code in the lock and doesn't keep an uninitialized variable around for a few lines.

Reply:
Actually none of that is necessary. `vChecks` doesn't change size during the adding loop. The old `vChecks.size()` based checks can be kept.

Reply:
>  `vChecks` doesn't change size during the adding loop.

Correct. But `vChecks` is passed in `CCheckQueue::Add` by a non-const reference, and that makes possible its modification by a concurrent thread between leaving a critical section and checking `num_of_new_checks == 1`.

Although, the only caller is located in `CChainState::ConnectBlock` which is guarded by the `cs_main` mutex. Therefore,  you are right.

> The old `vChecks.size()` based checks can be kept.

Yes, it will work. But while looking into `CCheckQueue::Add` code, it is not obvious that `vChecks.size()` is constant in multi-threading environment. I think that adding a local variable makes code more readable and robust considering any possible changes at the caller site(s) in the future.

OTOH, it seems all could look clearer if we pass `vChecks` as an rvalue reference using move semantics. What do you think?

Reply:
I would never think that a reference argument could be changed by another thread while the function is executing. I hope we don't do that anywhere in the code. Because how would the function avoid races? Which mutex protects the argument? That would require something like (it does not compile, of course):

```cpp
void func(T& arg GUARDED_BY(some_mutex), ...)
{
    ...
    LOCK(some_mutex);
    access arg;
    ...
```

So, I think it is safe to assume that `vChecks` will not change while `Add()` is executing. Indeed the only caller passes in a local variable that is not visible by other threads (and then throws it away, so the "swap" is actually a "move").

> ... it seems all could look clearer if we pass vChecks as an rvalue reference using move semantics ...

Yes :) and `std::move()` each element from `vChecks` into `queue`. It is unnecessary to first create a default-constructed object only to swap it with another and throw it away. This is what `std::move()` is for.

Reply:
@hebasto That's not how `const` works in C++. The lack thereof means *you* (the called function in this case) are allowed to modify the argument; it says nothing at all about whether another thread might be modifying the same variable. Even if the argument was `const`, nothing prevents another thread from holding a mutable reference to the same object.

That said, passing an argument (const or not) to a function while it is being modified by other threads would be completely non-idiomatic, dangerous code. It is possible, and sometimes necessary of course, but at least you'd expect the passed object to have internal mutexes or other synchronization mechanisms to protect its state in that case; not something the callee should deal with. Virtually every STL function will exhibit undefined behavior if you pass it references to arguments that are being modified concurrently. Outside of very exceptional cases that should raise big red flags, you can very much assume that arguments passed to a function are under exclusive control of that function for the duration of its execution.

Reply:
@sipa Thanks for your detailed explanation.

Going to address all of the comments.